### PR TITLE
Project Beats: generate sounds from sequenced samples

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Enables playback of generated sounds consisting of sequenced samples. This will be used to create sounds that can be automatically adjusted to any key or tempo.

Note: there is some [ongoing discussion](https://codedotorg.slack.com/archives/C04TH8400AU/p1685572925802829) on the exact format of the schema, which may affect this PR, but most of the logic and functionality should stay the same.

## Links

https://codedotorg.atlassian.net/browse/SL-887

## Testing story

Tested locally with a custom sequenced sample in a local manifest.

### CODE REVIEW

File: apps/src/music/utils/Notes.ts [hunk 1]
```diff
@@ -1,24 +1,42 @@
...
+// Used to calculate the specific project root note from the
+// project key. When calculating transposed notes, the note
+// offset is expressed as the number of semitones above the
+// project root note, starting at the octave of middle C
+// (MIDI note 60, or C4).
+//
+// For example, in the key of C (enum value 0), the root note
+// is 60 (0 + 60), or C4. In the key of F (enum value 6), the
+// root note is 64 (4 + 60), or F4.
+const ROOT_NOTE_OCTAVE = 60;
```
Comment (breville) [inline apps/src/music/utils/Notes.ts]: Just regarding the name, it seems like this is actually not an octave but just a note?

File: apps/src/music/player/soundSub.js [hunk 2]
```diff
@@ -108,23 +112,38 @@ WebAudio.prototype.PlaySoundByBuffer = function (
...
+  duration
 ) {
-  var source = audioContext.createBufferSource(); // creates a sound source
+  const source = audioContext.createBufferSource(); // creates a sound source
   source.buffer = audioBuffer; // tell the source which sound to play
+  let currentNode = source;
+
+  if (duration) {
+    // If playing for a specific duration, apply a small fadeout to the sound
+    // to prevent clicks and pops
```
Comment (breville) [inline apps/src/music/player/soundSub.js:124]: Is the click/pop the result of stopping a sound before it's reached the end of its sample?

File: apps/src/music/player/soundSub.js [hunk 3]
```diff
@@ -108,23 +112,38 @@ WebAudio.prototype.PlaySoundByBuffer = function (
...
+    const gainNode = audioContext.createGain();
+    gainNode.gain.setTargetAtTime(
+      0,
+      when + duration - RELEASE_DURATION_SECONDS,
+      RELEASE_TIME_CONSTANT
+    );
```
Comment (breville) [inline apps/src/music/player/soundSub.js:130]: Very cool functionality!

Just learning about this now.  Are there any interesting edge cases where we wouldn't get the full curve from full volume to 0, say if the `duration` is shorter than `RELEASE_DURATION_SECONDS `?

File: apps/src/music/utils/Notes.ts (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/music/utils/Notes.ts]: Yeah couldn't think of a good name for this value - maybe ROOT_NOTE is better then?

File: apps/src/music/player/soundSub.js (hunk 2 above)
Comment (sanchitmalhotra126) [reply apps/src/music/player/soundSub.js:124]: Yep exactly

File: apps/src/music/player/soundSub.js (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/music/player/soundSub.js:130]: Good question. For 120 BPM, this should be fine as 16th notes will be .125 seconds, but at higher tempos, a note duration could definitely be below the RELEASE_DURATION_SECONDS. This should probably be scaled based off of tempo.

File: apps/src/music/utils/Notes.ts (hunk 1 above)
Comment (breville) [reply apps/src/music/utils/Notes.ts]: I think that seems good.

File: apps/src/music/player/soundSub.js (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/music/player/soundSub.js:130]: I created a [task](https://codedotorg.atlassian.net/browse/SL-900) for supporting variable BPM, so I can tweak this as a part of that work.

File: apps/src/music/player/soundSub.js [hunk 4]
```diff
@@ -6,7 +6,11 @@ var audioContext = null;
 
 var soundEffects = null;
 
-// var soundSourceIdUpto = 0;
+// Length of time to fade out a sound, if trimming to a specific duration
+const RELEASE_DURATION_SECONDS = 0.1;
+// Time constant used to compute the release rate; at each time constant
+// interval the sound will decay exponentially.
+const RELEASE_TIME_CONSTANT = 0.075;
```
Comment (breville) [inline apps/src/music/player/soundSub.js:13]: Is `_CONSTANT` a naming convention for all constants, or just something specific to this concept?

Comment (sanchitmalhotra126) [reply apps/src/music/player/soundSub.js:13]: That's for this API specifically - https://developer.mozilla.org/en-US/docs/Web/API/AudioParam/setTargetAtTime

**Review by breville — APPROVED** [review]

**Pull request merged**